### ISSUE TITLE
init shouldn't need internet access

### ISSUE THREAD

**Issue by gamesguru** [openatx/uiautomator2#218]
Oftentimes I lose internet during a test and wonder what I would do if I needed to re-init a device or to.
I am not sure why the requirement, or why for example "apps can only be installed from http".  How hard would it be to also provide offline support?

Oftentimes it gets stuck here.  And it is not my internet, must be the server or something?  Either way makes me want an option where we download everything up front and just have it there, even if it is several megabytes.
`2018-09-18 07:29:52,377 - __main__.py:158 - INFO - app-uiautomator-test.apk downloading ...
`

Affects 0.1.3 and 0.1.4devs

**Comment by gamesguru** [comment]
That solves the install issue.

However if init is a requirement after simply rebooting the device, I am include to refactor the code in a way that **caches the downloads, or allows the user to manually place it in the cach**e and to avoid the online `http` process on that machine entirely.  If they are happy at a certain release, they should **be able to run** with that _in an offline environment_.

As for the lockup.  Well it didn't happen often and I'm not sure of the cause, but I'm just letting you know the download hung occasionally at the line mentioned in my original post.  Even if my internet connection or device connection died in the logic pool, still error message or timeout warning should be presented to end user..

**Comment by mingyuan-xia** [comment]
The download is not necessary for the innit after a device reboot. Basically, after reboot, you need an adb shell to restart all daemons (minitouch, minicap, then uiautomator2 server and atx-agent). I believe the init code can be trimmed a bit to only start thinks instead of checking/downloading latest daemons. PR is welcome, e.g., a new mode `--offline` for `init`

**Comment by gamesguru** [comment]
> the init code can be trimmed a bit to only start thinks instead of checking/downloading latest

I will submit one later on, thanks.

Question: I am not using this in a business setting anymore, but am happy to help your project out.  How are you and your friends using this, is it something that goes toe-to-toe with Appium or others in the corporate world, or more something that gets used at home?

**Comment by mingyuan-xia** [comment]
AFAIK, codeskyblue develops u2, atx-agent and atx-server based on inspiration from STF and xiaocong/uiautomator server. And he uses u2 for daily testing tasks in his corp and is kind enough to open it to a wider audience. So far we also have a Chinese community of over 600 ppl from different corps using u2 for UI automation scripts and atx-server for USB-less device farm management. The Chinese community is formed as a QQ group (IRC-like tool) while for English-speaking users, Github is for now.

I am integrating openatx tools into our product, [AppetizerIO](https://www.appetizer.io) to make these tools easier to install and use (stand-alone and cooperative with other tools).
A bit of ads here:
![2](https://user-images.githubusercontent.com/1170592/47591405-0ebd0c80-d9a2-11e8-8a46-484ff5cb5b15.gif)
![toolmgr](https://user-images.githubusercontent.com/1170592/47591425-26949080-d9a2-11e8-8984-38979183a5ab.gif)
The first is an enhanced Hierarchy Dump/Viewer to code up u2/appium scripts.
The second is to install/start/stop/remove tools in a dockerized manner.
Sorry, I recorded these use cases for Chinese language. We do provide English version.

**Comment by gamesguru** [comment]
That is interesting I will respond more when I feel. I actually came here cause I had another immediate question as it relates to the issue at hand:

With MyFire command line parser, are we able to specify multiple flags to one action, e.g. `--offline` and `--no-update` both also trigger the same thing?

**Comment by gamesguru** [comment]
I agree it can work with lots of other software, unit test, logging, whatever you need.  I like it, my boss didn't.

The problem I think is, does it would work in Test Cloud environment or remote testing (obviously depends if we have python2)?

**Comment by mingyuan-xia** [comment]
Regarding Fire, should be. After all, it is Google testified.
It would be nice to have fine-grained flags and please update README when you submit the PR.
Currently, we have over 40% issues just regarding init. So it would be the part to be improved most in the near future.

Test cloud/remote testing:
* By far, using u2 with devices from a test cloud via Internet is tricky because we need a public IP:port for uiautomator/atx-agent daemons running on the device. But I do hear people trying out in-premise atx-server-managed test cloud + u2.
* remote testing, Yes. Someone is trying out on-device QPython + u2 script shipped to the device. Personally I never tried. But sound like an interesting attempt.

**Comment by gamesguru** [comment]
Also, Fire is not perfect.

It is converting the serial 87384e7594 to floating point

csb and I already had to make one workaround with that. still need to update readme for it too

**Comment by mingyuan-xia** [comment]
@gamesguru Agent on the remote host should be perfectly supported, as long as the devices are properly connected with the remote host (USB/same WiFi access point). I used to have a travis task running the agent connected with the an Android emulator started there. Worked like a charm. For the remote host, python and adb are needed as you said.

**Comment by mingyuan-xia** [comment]
Also I have experience with click and the standard cmd, but no experience with Fire. I think a string type for serial is the best fit. As we might have a serial like "192.168.0.1:5555" for emulators.

**Comment by gamesguru** [comment]
anything like that is automatically a string.  84574e35845 is floating point according to Fire, string according to sys.argv

i suspect many remote hosts are not having python, and we will have to convince them of us.
